Ct 2109 extend sar deadline

mseedat-moj opened the pull request:
### Summary
Allow user to extend the deadline to respond to a SAR for up to 30 days at a time

### Ticket
Implements both add and remove SAR extension deadline due to ticket confusion.

https://dsdmoj.atlassian.net/browse/CT-2109
https://dsdmoj.atlassian.net/browse/CT-2116

### Notes

-  `extend_for_pit` implementation change can be found in PR #677 
- Assumes extension is done using Calendar days rather than Business days
- Feature test for Approver user extending a SAR deadline is outstanding (skipped test)
- Note the alteration in tests related to case history as the use of `simple_format` exposes
   line break characters stored in the database that were previously not retrieved

### Screenshots
![screen shot 2019-02-14 at 01 37 47](https://user-images.githubusercontent.com/47113046/52756426-35f32480-2ff9-11e9-957e-a93baff87017.png)

![screen shot 2019-02-14 at 01 37 58](https://user-images.githubusercontent.com/47113046/52756431-3be90580-2ff9-11e9-8f87-cc3b706132dc.png)

Commit:
CT-2109 Add new extend SAR deadline route

Commit:
CT-2109 Display extend deadline button in action bar

- TODO: Authorization policy is undefined

Commit:
CT-2109 Set custom days limit for SAR extensions

Commit:
CT-2109 Display deadline extension form

- Note: heavily inspired by PIT extension mechanism and form

Commit:
CTS-2109 Add new route to process SAR extension

Commit:
CT-2109 Execute SAR extension with success/failure message

Commit:
CT-2109 Test SAR extension routing and controller handling

beccapearce reviewed (commented):
Good start! Once we add the event to the state machine, we can look at adding it to the transition decorator so It'll show up in the case history, and adding it to out event tests.

Commit:
CT-2109 Replace arbitrary date field with restricted extension period

- Note: decision made to restrict the user to only select a fixed
  extension period

Commit:
CT-2109 Use submitted extension period (30/60 days) to set new extension date

Commit:
CT-2109 Remove unused HTML

Commit:
CT-2109 Clean up execute_extend_deadline_for_sar with use of translations

Commit:
CT-2109 Add new remove_extended_deadline_for_sar route

Commit:
CT-2109 New service to remove SAR extension, based on PIT extension removal

Commit:
CT-2109 Display 'remove extension' button in case action bar

Commit:
CT-2109 Execute removal of SAR extension deadline using dummy date

Commit:
CT-2109 Introduce #deadline_extended? property for Case::SAR models

Commit:
CT-2109 Add new SAR event extend_deadline_for_sar

Commit:
CT-2109 Remove explicit output of Extend Deadline button

- Note: permitted_events will automatically output the button as long
  as the button has a definition in case_helper.rb

Commit:
CT-2109 Generate state transition for deadline extension

Commit:
CT-2109 Set deadline_extended property

Commit:
CT-2109 Correctly retrieve original final deadline for SAR

Commit:
CT-2109 Alter executing service to perform extension calculation

- Also includes additional #message to explicitly specify the
  number of days the SAR is to be extended by

Commit:
CT-2109 Output SAR deadline extension reason/message in case history

Commit:
CT-2109 Amend test to reflect changes in CaseExtendedForSARService

Commit:
CT-2109 Add extend_deadline_for_sar event to state machine

- Based on discussions with Rebecca regarding appropriate rules and
  roles

Commit:
CT-2109 Improve formatting of validation logic

Commit:
CT-2109 Fix SAR deadline extension radio buttons

Commit:
CT-2109 Amend SAR deadline extension validation attribute name

Commit:
CT-2109 Add remove SAR deadline extension state machine events & predicates

Commit:
CT-2109 Temporarily remove Remove deadline button from action bar

- Relying on State Machine to output the button when possible

Commit:
CT-2109 Remove SAR deadline extension if remove action button pressed

Commit:
CT-2109 Change maximum SAR extension limit to 60 days

- Note: SARs can be extended for a total of 60 days. An extension can be
  2 x 30 days or 1 x 60 days from the original deadline date. The total
  SAR deadline period would then be up to 90 days (inc. original period)

Commit:
CT-2109 Add missing remove SAR deadline event translation

Commit:
CT-2109 Set correct class styles for event action buttons (blue buttons)

Commit:
CT-2109 Implement correct rules for SAR deadline extension

Commit:
CT-2109 Alphabetically order SAR deadline event in list of events

Commit:
CT-2109 Alphabetically order translation entries for SAR deadline extension

Commit:
CT-2109 Correctly define authorization rules for SAR extension/removal

Commit:
CT-2109 Alphabetically arrange SAR deadline events in cases_controller

Commit:
CT-2109 Extend SAR deadlines using CALENDAR days not working

Commit:
CT-2109 Amend SAR extension output to reflect design

Commit:
CT-2109 Ensure validation uses calendar days not working days

Commit:
CT-2109 Amend SAR deadline extension form to reflect design

Commit:
CT-2109 Replace generic team with correct Case Team value for state machine

Commit:
CT-2109 Amend messages_controller to use User#sort_teams_by_roles

Commit:
CT-2109 Rename state 'remove_extended_deadline_for_sar'

Commit:
CT-2109 Rename state 'extend_deadline_for_sar'

- Includes re-generated routes list in routes.rb

Commit:
CT-2109 Rename Service classes to use new state name

Commit:
CT-2109 Add new ExtendSARDeadline SitePrism Page for testing

Commit:
CT-2109 Add feature test for SAR extension by a manager

Commit:
CT-2109 Update feature test by extracting common steps

Commit:
CT-2109 Amend feature to test for removal of extension deadline

Commit:
CT-2109 Add missing translation for state machine error condition

Commit:
CT-2109 Add feature test for 60 day SAR extension

Commit:
CT-2109 Add feature test for responder trying to extend a SAR

Commit:
CT-2109 Add cases_controller extend_sar_deadline tests

Commit:
CT-2109 Rename decorator to CaseExtendSARDeadlineDecorator

Commit:
CT-2109 Specify correct model to decorate (SAR)

Commit:
CT-2109 Test CaseExtendSARDeadlineDecorator class

Commit:
CT-2109 Rename translation key to be more succinct

Commit:
CT-2109 Add test suite for CaseExtendSARDeadlineService

Commit:
CT-2109 Extract common value for CaseExtendSARDeadlineService spec

Commit:
CT-2109 Fix factory for a SAR with extended deadline

Commit:
CT-2109 Add CaseRemoveSARDeadlineExtensionService spec

Commit:
CT-2109 Rename spec variable name to 'initial_deadline'

Commit:
CT-2109 Add retrospective SAR deadline extension test

Commit:
CT-2109 Add retrospective SAR deadline extension removal test

Commit:
CT-2109 Add basic SAR Pundit Policy spec

Commit:
CT-2109 Rename SAR::max_allowed_extension_date to max_allowed_deadline_date

Commit:
CT-2109 Add basic SAR model deadline spec

Commit:
CT-2109 Add additional SAR model spec tests

Commit:
CT-2109 Add spec for extend SAR deadline actions on case show page

Commit:
CT-2109 Add pending test for Approver user extending SAR deadline

Commit:
CT-2109 Remove outdated TODO message

Commit:
CT-2109 Fix rubocop spacing error in spec

Commit:
CT-2109 Fix rubocop offense - too many method lines

- Note: addition of new events in show.html.slim spec caused this offense

Commit:
CT-2109 Fix case history string check feature spec errors

Commit:
CT-2109 Fix spec errors with incorrect SAR case type and string check

Commit:
CT-2109 Add missing events to spec

Commit:
CT-2109 Add missing events to spec

beccapearce reviewed (changes requested):
Obviously some tests still to fix. 
Just an overview of what I spotted since it seems there's more work to do on it to get the tests passing

Commit:
CT-2109 Remove unused translations

Commit:
CT-2109 Fix spelling error 'extention'

Commit:
CT-2109 Improve function name in specs

Commit:
CT-2109 Replace multi-line ternary with explicit If statement

Commit:
CT-2109 Replace hard-coded strings with translation

Commit:
CT-2109 Remove unnecessary comment

Commit:
CT-2109 Move page title translations into dedicated page_titles file

Commit:
CT-2109 Add missing translation

Commit:
CT-2109 Further improvement of cases_controller.rb SAR extension methods

Commit:
CT-2109 Amend code as per PR review recommendations

- rename method Case::SAR#extendable? to deadline_extendable?
- remove reloads in Case::SAR#extend_deadline! and reset_deadline!
- User#sort_teams_by_roles now accepts custom role weightings
- Add named parameter to case_extend_sar_deadline_service_spec

Commit:
CT-2109 Fix broken test due to change in controller execution

Commit:
CT-2109 Rename predicate methods to be more descriptive of implementation

Commit:
CT-2109 Enable skipped test with correct approver user setup

Commit:
CT-2109 Replace un-named rspec subject with named subject

Commit:
CT-2109 Add note to clarify role weightings can be arbitrary

Commit:
CT-2109 Rename and change predicate for approver removing SAR deadline

Commit:
CT-2109 Add new pit_deadline_property to all Case types

- replaces explicit checking for historical case transitions and instead
  relies on a flag (has_pit_extension)

Commit:
CT-2109 Improve formatting for SAR#initial_deadline

Commit:
CT-2109 Replace hard-coded team with actual team for case

Commit:
CT-2109 Fix failing tests by using correct case property name

Commit:
CT-2109 Update SQL Structure after migration

Commit:
CT-2109 Skip test that may be incorrect (TODO)

Commit:
CT-2109 Fix rubocop error

Commit:
CT-2109 Hard code acting_team for PIT extension/removal

- Note: Without using 'acting_team: BusinessUnit.dacu_bmt' a significant
  number of existing tests fail which is beyond the scope of this work.
  This indicates either incorrectly setup factory objects, misuse of teams
  or an issue with the services.

Commit:
CT-2019 Amend test because of changed ordering of output

Commit:
CT-2109 Remove leftover puts statement

Commit:
CT-2109 Add explanatory note for predicate use of defensive 'try'

Commit:
CT-2109 Re-write validation of parameters in CaseExtendSARDeadlineService

Commit:
CT-2109 Rearrange test steps to make test intention clearer

Commit:
CT-2109 Add parameterized service constructor

- Specs changed accordingly
- Specs re-evaluated for use of subject and exclamation let/subject

Commit:
CT-2109 Use named parameters for step function

Commit:
CT-2109 Fix new rubocop errors

Commit:
CT-2109 Fix broken test due to introduction of named parameters

Commit:
CT-2109 Replace factory with sar extension trait

Commit:
CT-2109 Improve remove_sar_deadline_service_spec

Commit:
CT-2109 Replace schema migration with data-migration

Commit:
CT-2109 Force ordering of test output to prevent intermittent test failure
